ydb/src/load_balancer/nearest_dc_balancer.rs

```diff
@@ -0,0 +1,388 @@
+        Ok(())
+    }
+
+    fn waiter(&self) -> Box<dyn Waiter> {
+        let self_waiter = Box::new(self.waiter.clone());
+        match self.config.fallback_strategy.borrow() {
+            FallbackStrategy::BalanceWithOther(balancer) => {
+                Box::new(AllWaiter::new(vec![self_waiter, balancer.waiter()]))
+            }
+            FallbackStrategy::Error => self_waiter,
+        }
+    }
+}
+
+pub(super) const NODES_PER_DC: usize = 5;
+pub(super) const PING_TIMEOUT_SECS: u64 = 60;
+
+impl NearestDCBalancer {
+    fn get_endpoint(&self, service: Service) -> YdbResult<Uri> {
+        match self.balancer_state.try_lock() {
+            Ok(state_guard) => {
+                match state_guard
+                    .borrow()
+                    .preferred_endpoints
+                    .choose(&mut thread_rng())
+                {
+                    Some(ep) => return YdbResult::Ok(ep.uri.clone()),
+                    None => (),
+                }
+                match self.config.fallback_strategy.borrow() {
+                    FallbackStrategy::Error => Err(YdbError::custom(format!(
+                        "no available endpoints for service:{}",
+                        service
+                    ))),
+                    FallbackStrategy::BalanceWithOther(balancer) => {
+                        info!("trying fallback balancer...");
+                        balancer.endpoint(service)
+                    }
+                }
+            }
+            Err(_) => Err(YdbError::Custom(
+                "balancer is updating its state".to_string(),
+            )),
+        }
+    }
+
+    pub(super) async fn adjust_local_dc(
+        balancer_state: Arc<Mutex<BalancerState>>,
+        mut state_reciever: watch::Receiver<Arc<DiscoveryState>>,
+        stop_ping_process: CancellationToken,
+        waiter: Arc<WaiterImpl>,
+    ) {
+        loop {
+            tokio::select! {
+                _ = stop_ping_process.cancelled() => {
+                    return
+                }
+                result = state_reciever.changed() =>{
+                    if result.is_err(){ // sender have been dropped
+                        return
+                    }
+                }
+            }
+            let new_discovery_state = state_reciever.borrow_and_update().clone();
+            match Self::extract_nodes(&new_discovery_state) {
+                Ok(some_nodes) => {
+                    let mut dc_to_nodes = Self::split_endpoints_by_location(some_nodes);
+                    let mut to_check = Vec::with_capacity(NODES_PER_DC * dc_to_nodes.keys().len());
+                    dc_to_nodes.values_mut().for_each(|endpoints| {
+                        to_check.append(Self::get_random_endpoints(endpoints))
+                    });
+                    match Self::find_local_dc(&to_check).await {
+                        Ok(dc) => {
+                            info!("found new local dc:{}", dc);
```

Comment:
it is often message, not intresting when all ok. Better to use debug for the message